File: 20.Concurrency & Parallelism - IO Bound vs CPU Bound/code/typescript/a_typical_go_handler_2.ts
import express from 'express';
import { Pool } from 'pg';

const app = express();
const db = new Pool();

app.get('/user', async (req, res) => {
    const userID = req.query.id;
    
    try {
        // This does NOT block the main thread. It yields back to the Event Loop.
        const result = await db.query('SELECT name, email FROM users WHERE id = $1', [userID]);
        if (result.rows.length === 0) {
            return res.status(404).send('not found');
        }
        // Resumes here after DB responds
        res.json(result.rows[0]);
    } catch (err) {
        res.status(500).send('error');
    }
});

File: 20.Concurrency & Parallelism - IO Bound vs CPU Bound/code/rust/a_typical_go_handler_2.rs
use actix_web::{web, HttpResponse, Responder};
use sqlx::PgPool;
use serde::Serialize;

#[derive(Serialize)]
struct User { name: String, email: String }

async fn handle_get_user(db: web::Data<PgPool>, query: web::Query<std::collections::HashMap<String, String>>) -> impl Responder {
    let user_id = query.get("id").unwrap();

    // This BLOCKS the task, but not the OS thread (yields back to Tokio reactor).
    match sqlx::query_as!(User, "SELECT name, email FROM users WHERE id = $1", user_id.parse::<i64>().unwrap_or(0)).fetch_optional(db.get_ref()).await {
        Ok(Some(user)) => HttpResponse::Ok().json(user),
        Ok(None) => HttpResponse::NotFound().body("not found"),
        Err(_) => HttpResponse::InternalServerError().finish(),
    }
}
